pkg/kubelet/cm/deviceplugin/pod_devices.go

```diff
@@ -17,207 +17,346 @@ limitations under the License.
 package deviceplugin
 
 import (
+	"sync"
+
 	"github.com/golang/glog"
 
+	"k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/util/sets"
 	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha"
 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
 )
 
-type deviceAllocateInfo struct {
-	// deviceIds contains device Ids allocated to this container for the given resourceName.
-	deviceIds sets.String
+type devAllocation interface {
+	// get a deep copy of all maps managed.
+	allocations() map[string]containerMap
+
+	// get the set of pod IDs known to the device manager
+	pods() sets.String
+
+	// Check if the specified pod uid is known to have allocations
+	hasPod(pod string) bool
+
+	// Enumerate all devices known to be already allocated by returning a map
+	// keyed by resource name.
+	devices() map[string]sets.String
+
+	// Enumerate all known device allocations for the given container on
+	// specified resource.
+	devicesByContainer(pod, container, resource string) sets.String
+
+	// Insert new allocation record to the registry.
+	insert(pod, container, resource string, devices sets.String, resp *pluginapi.AllocateResponse)
+
+	// Update the allocation registry based on the provided list of active
+	// pods. Records for inactive pods are pruned.
+	update(activePods []*v1.Pod) map[string]sets.String
+
+	// Marshal the registry into data for checkpointing
+	toCheckpointData() []podDevicesCheckpointEntry
+
+	// Unmarshal checkpointed data into registry
+	fromCheckpointData(data []podDevicesCheckpointEntry)
+
+	// Collect device information into container runtime settings.
+	collectRunOptions(pod, container string) *DeviceRunContainerOptions
+}
+
+type deviceInfo struct {
+	// ids contains device Ids allocated to this container for the given resourceName.
+	ids sets.String
 	// allocResp contains cached rpc AllocateResponse.
 	allocResp *pluginapi.AllocateResponse
 }
 
-type resourceAllocateInfo map[string]deviceAllocateInfo // Keyed by resourceName.
-type containerDevices map[string]resourceAllocateInfo   // Keyed by containerName.
-type podDevices map[string]containerDevices             // Keyed by podUID.
+type resourceMap map[string]deviceInfo   // Keyed by resourceName.
+type containerMap map[string]resourceMap // Keyed by containerName.
+
+// devAllocationImpl is a struct that implements the allocation interface.
+type devAllocationImpl struct {
+	sync.RWMutex
```

Comment:
I'm wondering if a simple Mutex might be more appropriate here.

From what I remember (though I might be wrong) the allocate function is always called from a single go routine.
So we have one reader and one writer. Which is not really the use case the RWMutex is expected to be used for :)

Reply:
Thanks for pointing this out. In PRs that follow up, I will pay attention to the invocations of all these interface functions. If RWMutex is not buying us anything, I'll switch it back to Mutex for simplicity.